### ISSUE TITLE
Last two updates has gobbled up memory

### ISSUE THREAD

**Issue by TanoopSungha** [potmat/homebridge-google-nest-sdm#150]
After the last two updates I now find that it now gobbles up RAM on the Pi. When I restart the child bridge it then resolves the issue. I am using a Pi5.

**Comment by potmat** [comment]
That's very odd, I can't account for it as there were no changes to the plugin itself, merely the arguments is passes to ffmpeg.  Can you see which process is using the RAM (Homebridge or ffmpeg)?

**Comment by TanoopSungha** [comment]
The memory only goes down when I report or restart the child bridge to google nest sim

**Comment by potmat** [comment]
Hi @TanoopSungha, I've not be able to reproduce this, my Pi4 still has 95% memory available after weeks of uptime.  I'll leave it open if anyone can reproduce it and submit a PR.

**Comment by hillaker** [comment]
@potmat, I am seeing the same behavior as TanoopSungha on my Raspberry Pi 5-- memory slowly disappears over the course of the day until I have less than 0.5GB available (out of 8). 

Please find a screenshot of top sorted by memory-- it looks like ffmpeg is sticking around after a particular recording is complete. Happy to share anything else that might help identify a fix-- I'm running Kali linux, rather than Raspian or similar, and it's the ARM64 version. Not sure if that's relevant for this issue: 

![image](https://github.com/potmat/homebridge-google-nest-sdm/assets/5393987/f73de082-f457-4dbe-82b4-22e5601e8b45)

Memory after letting the plugin run for ~9 hours (after opening a stream, I left it running without opening a stream earlier in the day and didn't notice an increase in memory usage: 
<img width="411" alt="image" src="https://github.com/potmat/homebridge-google-nest-sdm/assets/5393987/c4dde5cd-cb9c-40d7-a803-7a797493a2b1">

Memory after restarting the plugin's bridge (but not home bridge), taken ~1 minute after the previous screenshot: 
<img width="404" alt="image" src="https://github.com/potmat/homebridge-google-nest-sdm/assets/5393987/7c2bf310-eae1-47d4-9b12-198ae1636629">

If there's something I can do to change my configuration to make it easier to diagnose the issue, please let me know and I'll give it a shot and share the results. Thanks!

Edited again to add that I only have this plugin as one that an outputs video-- my other plugins are switches.

**Comment by GillesWi** [comment]
Yes, same behaviour on Raspberry Pi 4 (4Gb). After 7 days the memory increases so much that Homebridge just crashes and I need to unplug the Rpi to get it working again. If I restart the plugin within those days there are no problems...

And a week ago I lost my whole homebridge instance, everything was corrupt, so be sure to backup in time ;)!

**Comment by ShadowRegent** [comment]
Can confirm that I'm also seeing orphaned ffmpeg processes consuming high amounts of memory.  I had 16 processes with just 5 hours of uptime.  Killing them (with SIGKILL, they were unresponsive to SIGTERM) released memory and left the following logs after killing:

`[homebridge-google-nest-sdm] Encountered unexpected error on generator Error: FFMPEG socket closed during read for 8 bytes!
    at Socket.endHandler (/usr/local/lib/node_modules/homebridge-google-nest-sdm/src/HksvStreamer.ts:190:24)
    at Socket.emit (node:events:519:28)
    at TCP.<anonymous> (node:net:338:12)`

This was preceded by many errors that looked like this:

`Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.`

**Comment by ShadowRegent** [comment]
Was anyone else who experienced this using "copy" as vEncoder?  I switched back to default and I haven't seen the issue recur (I need to give it a bit more time to be sure though).

**Comment by potmat** [comment]
@ShadowRegent:  Ok that means that HKSV is what's causing the issue.

It's starting up streams to record motion (or whatever) events and the FFMpeg process are not getting properly cleaned up.

To everyone in the thread, turning off HKSV will likely solve the issue.  I know that's not great, but we can continue to investigate HKSV and hopefully find out why so many processes are being orphaned.

**Comment by JulesTKirk** [comment]
Chiming in, it also seems that I am having issues with HKSV or something related to FFMpeg : 

```
[8/11/2024, 2:14:32 AM] [homebridge-google-nest-sdm] Encountered unexpected error on generator Error: FFMPEG socket closed during read for 8 bytes!
    at Socket.endHandler (/volume1/homebridge/node_modules/homebridge-google-nest-sdm/src/HksvStreamer.ts:190:24)
    at Socket.emit (node:events:514:28)
    at TCP.<anonymous> (node:net:337:12)
```

which was preceeded by a high volume of these errors : 

```
Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
[8/9/2024, 12:02:21 PM] [homebridge-google-nest-sdm] Encountered unexpected error on generator Error: Unexpected state!
    at HksvStreamer.generator (/volume1/homebridge/node_modules/homebridge-google-nest-sdm/src/HksvStreamer.ts:141:19)
    at DoorbellStreamingDelegate.handleRecordingStreamRequest (/volume1/homebridge/node_modules/homebridge-google-nest-sdm/src/StreamingDelegate.ts:559:24)
```

I'm running on a Synology DS418 NAS with about 2GB of RAM so it was filled pretty fast and left the system completely unresponsive :(

**Comment by gonzalorodrigo** [comment]
Observing this issue too, zombie ffmpeg processes when recording happens. They accumulate consuming memory until the OOM killer starts to kick in, sometimes killing core processes of the system.

- Homebridge Google Nest Sdm `v1.1.23`
- ffmpeg-for-homebridge: `2.1.1`
- Platform RPi 5 with Debian: `Linux raspberrypi 6.6.51+rpt-rpi-2712 #1 SMP PREEMPT Debian 1:6.6.51-1+rpt3 (2024-10-08) aarch64 GNU/Linux`
- Suspicious log entries:
```log
[31/12/2024, 17:50:09] [homebridge-google-nest-sdm] Encountered unexpected error on generator Error: Unexpected state!
    at HksvStreamer.generator (/var/lib/homebridge/node_modules/homebridge-google-nest-sdm/src/HksvStreamer.ts:141:19)
    at CameraStreamingDelegate.handleRecordingStreamRequest (/var/lib/homebridge/node_modules/homebridge-google-nest-sdm/src/StreamingDelegate.ts:559:24)
[HDS ::ffff:192.168.4.215] Delegate finished recording stream 1 without sending anything out.       Controller will CANCEL.
[HDS ::ffff:192.168.4.215] Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
[HDS ::ffff:192.168.4.215] Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
[HDS ::ffff:192.168.4.215] Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
[HDS ::ffff:192.168.4.215] Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
[HDS ::ffff:192.168.4.215] Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
[HDS ::ffff:192.168.4.215] Recording download stream 1 is still awaiting generator although stream was closed 10s ago! This is a programming mistake by the camera implementation which prevents freeing up resources.
```


Looking a bit, one can see that the ffmpeg process have connections in CLOSE_WAIT with data still missing to be flushed:
```bash
$ sudo netstat -p  | grep ffmpeg
tcp   109266      0 192.168.5.124:34216     77.67.201.35.bc.g:https CLOSE_WAIT  79060/ffmpeg        
tcp        0      0 localhost:34510         localhost:39467         ESTABLISHED 79298/ffmpeg        
tcp   107885      0 192.168.5.124:57266     77.67.201.35.bc.g:https CLOSE_WAIT  79298/ffmpeg        
tcp   108462      0 192.168.5.124:44836     77.67.201.35.bc.g:https CLOSE_WAIT  78948/ffmpeg        
tcp        0      0 localhost:42506         localhost:42677         ESTABLISHED 79060/ffmpeg        
tcp   111218      0 192.168.5.124:41704     77.67.201.35.bc.g:https CLOSE_WAIT  79446/ffmpeg        
tcp        0      0 localhost:43584         localhost:41061         ESTABLISHED 80595/ffmpeg        
tcp        0      0 localhost:38810         localhost:39601         ESTABLISHED 78948/ffmpeg        
tcp        0      0 localhost:39810         localhost:38643         ESTABLISHED 79174/ffmpeg        
tcp   101745      0 192.168.5.124:33920     77.67.201.35.bc.g:https CLOSE_WAIT  79174/ffmpeg        
tcp        0      0 localhost:49828         localhost:44789         ESTABLISHED 79446/ffmpeg        
tcp   120033      0 192.168.5.124:45690     77.67.201.35.bc.g:https CLOSE_WAIT  80595/ffmpeg        
```
Server processes that can't finish cleanly because of a socket with data pending to be sent to the client are a [classic systems problem](https://web.archive.org/web/20170113135705/http://unix.derkeiler.com/Mailing-Lists/SunManagers/2006-01/msg00367.html).  

When a process is in this state, SIGTERM will not kill it, it's stuck waiting on the socket close.

I am not sure if this is something to be solved in ffmpeg itself or if the plug-in should learn to identify this idle ffmpegs and kill them. Meanwhile, as a mitigation, I put this is root's cron to clean the ffmpeg processes that are zombies:


```crontab
*/5 * * * * bash -c 'netstat -p  | grep ffmpeg | grep CLOSE_WAIT | tr -s " " | cut -d " " -f7 | cut -d "/" -f1 | xargs -Ixxx kill -9  xxx' &> /tmp/ffmpeg_killer.log
```

So far no more crashes.

**Comment by potmat** [comment]
@gonzalorodrigo: Very informative thank you!  If SIGTERM will not help, what about SIGKILL?

**Comment by ShadowRegent** [comment]
SIGKILL works, I have a similar cron to kill off orphaned processes that I've been using for a few months.

**Comment by ajplotkin** [comment]
I think I found at least part of the mechanism here — it's a session-lifecycle bug in the recording handler, not something environmental.

`handleRecordingStreamRequest` sets `this.recordingSessionInfo = { hksvStreamer, nestStreamer }` without first tearing down a prior session. `recordingSessionInfo` is a single property, so if the hub starts a new recording before the previous one's close event fires (a brief reconnect, or rapid re-triggering under sustained motion), the previous `HksvStreamer` — and the ffmpeg it spawned — is orphaned: nothing references it to `destroy()` it anymore, and the previous session's late `closeRecordingStream` tears down the **new** session instead. The leaked ffmpeg keeps running, and they stack up → the memory growth you're seeing.

Opened #217 with a small fix (destroy the prior session before overwriting it, mirroring `closeRecordingStream`'s teardown). To confirm on your side: `ps aux | grep ffmpeg` after a burst of recordings should show the orphans before the fix and none after.

**Comment by potmat** [comment]
If anyone is experiencing this please try the latest release and advise.

**Comment by littlepope81** [comment]
@potmat — before anyone reports back on this, a heads-up that will otherwise produce misleading results: **v1.2.24 doesn't appear to have been published to npm.**

```
$ npm view homebridge-google-nest-sdm dist-tags
{ latest: '1.1.23', beta: '1.1.0-beta.1' }

$ npm view homebridge-google-nest-sdm@1.1.24 version
npm error code E404
npm error 404 No match found for version 1.1.24
```

npm's `latest` is still `1.1.23`, published 2024-02-21. The GitHub release `v1.2.24` went out 2026-07-21T23:24Z, but the registry hasn't got it.

That matters for this thread specifically: anyone who "tries the latest release" through the Homebridge UI plugin manager will be installing **1.1.23** — which predates #150's fix (08c4185), #217, and #223. They'd still see orphaned ffmpeg processes and report the bug as unfixed, and you'd be chasing something you've already closed.

(Also worth noting the tag and the package version differ — the tag is `v1.2.24` but `package.json` in it says `1.1.24`. Not a problem in itself, just easy to trip over when checking whether a publish landed.)

I'm testing the released code directly from the tag rather than npm, and will report the orphaned-ffmpeg results here separately.

**Comment by littlepope81** [comment]
Caught this red-handed on my bridge this morning while preparing to test v1.2.24, so here is a concrete reproduction with the memory cost quantified.

## Before (pre-fix code)

```
$ date
Wed Jul 22 08:28:58 EDT 2026

$ ps -eo pid,etime,rss,stat,args | grep ffmpeg
  PID  ELAPSED       RSS STAT COMMAND
  817  14:53:05   550260 Sl  .../ffmpeg -protocol_whitelist pipe,crypto,udp,rtp,fd
                              -analyzeduration 15000000 -probesize 100000000 -fpsprobesize 0 -i -
                              -acodec libfdk_aac -profile:a aac_eld -ar 48k -b:a 64k -ac 1 -f mp4 ...
16539  10:25:35   594736 Sl  (identical command line)

total RSS of the two: 1,144,996 KB  ≈ 1.09 GB
```

Two ffmpeg processes, **14h53m** and **10h25m** old, still resident, holding **~1.09 GB** between them. No live view or recording was active — the last live session had ended 25 minutes earlier. The `-acodec libfdk_aac -profile:a aac_eld ... -f mp4` output identifies these as the **HKSV recording** pipeline rather than live view, which matches the title of this issue exactly.

The mechanism is the one #150 and #217/#223 address: `handleRecordingStreamRequest` assigned

```js
this.recordingSessionInfo = { streamId, hksvStreamer, nestStreamer }
```

over the previous session **without tearing it down**, so when a hub started a new recording before the old close fired, the previous `HksvStreamer` and its ffmpeg child were orphaned with nothing left holding a reference. Confirmed absent from the build that produced the above: the "Tear down any prior recording session" block and the stale-session guard (grep count 0 for both).

**Disclosure on that build:** it was my own branch off the pre-fix base, not stock 1.1.23. It adds snapshot/bitrate features, and relevant here, it sets `STOP_AFTER_MOTION_STOP = true`, so recordings end on motion-stop rather than running until the controller closes them. That means recordings cycle *more often* than stock, which plausibly makes orphaning easier to hit — so treat the frequency as suggestive, not as a stock-1.1.23 rate. The teardown path itself is byte-for-byte the pre-fix logic, so the mechanism is the same one.

## After (v1.2.24)

Now running stock v1.2.24, installed from the tag (see my note above about npm still serving 1.1.23):

```
version 1.1.24
"Tear down any prior recording session"  -> present
"stale/replaced session" guard           -> present
snapshot/fork features                   -> absent (confirmed stock)
```

Post-restart: 10/10 accessories restored, 0 errors, **0 ffmpeg processes**. Both old orphans died with the restart, so the window starts from a clean slate.

I have a sampler logging the ffmpeg count and oldest-process age every 60s. I'll let it run through a day of real motion and report the trend here — a flat count over ten minutes would prove nothing about a leak that takes hours to show, which is why the before numbers above are ages rather than counts.

**Comment by ajplotkin** [comment]
Thanks for the quantified repro — two multi-hour-old orphans at ~1 GB is a clear picture, and grep-0 on the teardown block + stale-session guard pins it to exactly the pre-fix path.

Confirmed the npm point here too: `latest` is still 1.1.23, and neither 1.1.24 nor 1.2.24 is on the registry, so a Homebridge-UI install pulls the pre-fix code. That's a publish only @potmat can do — flagging so the "tried latest, still leaking" reports don't come back as false negatives against an already-merged fix.

**Comment by littlepope81** [comment]
Follow-up to my reproduction above, now with the v1.2.24 result.

## Observation window

**08:33:56 → 13:42:24 EDT (5h 08m), 309 samples at 60s intervals**, stock v1.2.24 (installed from the tag), x86 docker-homebridge, 6 cameras, 5 of them WebRTC.

## The test was actually exercised

Worth establishing first, because a flat count on an idle box would prove nothing:

```
mp4 box lines (HKSV fragments written)  : 211
recording session start/close markers   : 12
motion events                           : 32
live view sessions                      : 9
```

So HKSV recordings genuinely ran during the window — this isn't an idle-box result.

## ffmpeg process counts

```
samples at count=0 : 296  (95.8%)
samples at count=1 : 11
samples at count=2 : 2
maximum concurrent : 2

every non-zero sample, with the oldest process age at that moment:
08:37:56 count=1 oldest=00:22     09:27:56 count=1 oldest=00:16
08:38:56 count=1 oldest=01:22     09:49:57 count=1 oldest=00:01
08:39:56 count=1 oldest=02:22     11:06:58 count=1 oldest=00:01
08:40:56 count=2 oldest=00:32     11:10:58 count=1 oldest=00:05
08:41:56 count=1 oldest=01:17     11:23:58 count=2 oldest=00:20
08:42:56 count=1 oldest=02:17     12:58:00 count=1 oldest=00:33
08:43:56 count=1 oldest=03:18
```

**Maximum age reached by any ffmpeg process: 3m 18s.** Every non-zero reading returned to zero on the following samples. Nothing survived past the length of the recording that spawned it.

## Before vs after

| | pre-fix build | v1.2.24 |
|---|---|---|
| oldest ffmpeg observed | **14h 53m** / **10h 25m** | **3m 18s** |
| resident memory held | ~1.09 GB | returns to 0 |
| processes at idle | 2, indefinitely | 0 |

## What this does and doesn't establish

It **does** show that under 12 recording sessions across 5 hours, no ffmpeg child outlived its recording, and the count returned to zero every time. Under the pre-fix code on the same hardware the previous evening, two did outlive theirs and were still resident 10–15 hours later. That is a clear behavioural difference on the exact path this issue is about.

It **doesn't** amount to a multi-day soak. The orphan needs a specific race — a hub starting a new recording before the previous close lands — and 12 sessions is not a large sample of chances to hit it. I'd call this a strong positive signal rather than proof of absence. I'll leave the sampler running and will report back if anything ages unexpectedly.

## Other observations

No plugin-level errors in the window (uppercase `ERROR` count: 0). There are ~150 lines matching "error" case-insensitively, but they're all FFmpeg's own decoder chatter passed through at debug level — `[h264 @ ...] error while decoding MB ...` and `[sdp @ ...] [error]` — which is ordinary packet-loss noise on WebRTC input, not plugin failures. Flagging it only because a naive `grep -i error` on this log looks alarming and isn't.

Camera streams start cleanly on v1.2.24: probe complete around 1.1s, first keyframe fully received around 1.6s, REMB going out ~2ms after the first RTP packet.

**Comment by potmat** [comment]
Sorry, should be published to NPM now.

**Comment by littlepope81** [comment]
Confirmed published and the artifact is good — verified rather than assumed, since a publish cut from a stale tree would be worse than no publish at all:

```
$ npm view homebridge-google-nest-sdm dist-tags.latest
1.1.24

$ npm pack homebridge-google-nest-sdm@1.1.24   # then, in the extracted package/
"Tear down any prior recording session"  -> present in dist/StreamingDelegate.js
"stale/replaced session" guard           -> present in dist/StreamingDelegate.js
```

And the published `dist/` is **byte-identical** to the `v1.2.24` tag for `StreamingDelegate.js`, `NestStreamer.js` and `sdm/Api.js` — so what's on the registry is what's in the tag, not a stale local build.

That means anyone landing on this thread can now just update through the Homebridge UI and actually get the fix, which wasn't true a few hours ago. My orphan-test results above were run against this same code installed from the tag, so they carry over to the npm build unchanged.

Thanks for turning that around quickly.

**Comment by ajplotkin** [comment]
Great soak — thanks for quantifying it. 3m18s max age vs the 14–15h orphans on the pre-fix build, returning to zero across 12 sessions, is exactly the behavioral change we were after: each session now tears down on its own exit instead of waiting for the next recording (or never).

Agreed it's a strong signal rather than proof of absence — the race needs the hub to overlap two recordings, and 12 sessions isn't many chances at it — so leaving the sampler running is the right call.

And good flag on the error-grep noise: the `[h264 …] error while decoding MB` lines are WebRTC packet-loss chatter, not plugin failures — easy to misread on a first grep.

**Comment by littlepope81** [comment]
10-hour update on the v1.2.24 soak.

## Window

**08:33:56 → 18:53:05 EDT — 10h 19m, 620 samples at 60s.** Same box, stock v1.2.24, 6 cameras.

```
samples at count=0 : 599  (96.6%)
samples at count=1 : 18
samples at count=2 : 3
maximum concurrent : 2
maximum age reached by any ffmpeg process : 3m 18s
ffmpeg processes at the time of writing   : 0
```

Every non-zero reading returned to zero within a sample or two. The 3m18s peak is still the one from the very first cluster at 08:43 — nothing in the following nine and a half hours got close to it.

Non-zero samples after the 5h checkpoint, for completeness:

```
14:08 count=1 oldest=00:14    17:04 count=1 oldest=00:15
14:51 count=1 oldest=00:13    17:35 count=2 oldest=00:23
16:34 count=1 oldest=00:02    18:16 count=1 oldest=00:19
                              18:21 count=1 oldest=00:02
                              18:43 count=1 oldest=00:30
```

## Still exercised, not idle

The plugin log rotated partway through, so its counters are **not** cumulative and I'm not going to add them up. What the current (afternoon-onward) log shows on its own:

```
mp4 box lines written : 171   (all of them PM — recording continued through the afternoon)
recording markers     : 14
motion events         : 34
uppercase ERROR lines : 0
```

So recordings kept running for the whole window, not just the morning.

## The comparison that matters

An orphan created when this window opened would, right now, be **10h 19m old and still resident** — which is precisely the state the two processes I found on the pre-fix build were in (14h53m and 10h25m, ~1.09 GB between them). Nothing on this build survived past 3m18s.

I'd now call this more than a strong signal, though still short of proof: a full day of ordinary use, recordings throughout, and the failure mode simply doesn't appear. The race still needs overlapping recordings and I can't guarantee how many chances it got. I'll leave the sampler running and will report back only if something contradicts this — silence from me should be read as "still clean."

**Comment by ajplotkin** [comment]
This still reproduces, and I have two open PRs that I believe address the mechanism — flagging rather than claiming, since I have not reproduced your exact stack trace.

**The orphans are real and I have counts.** On a Pi 4 I found three orphaned ffmpeg processes alive simultaneously, one of them for **6.3 hours**, each still holding an open input. They ignore `SIGTERM`: an ffmpeg blocked reading an input that never delivers will not exit on a term signal, which matches the "only SIGKILL clears them" reports above.

Two distinct causes, both fixable in this plugin:

1. **The recording session is never torn down on some exit paths.** [#224](https://github.com/potmat/homebridge-google-nest-sdm/pull/224) moves the reset into a `finally`, so an exception between `start()` and the generator finishing cannot leave `recordingSessionInfo` populated and its ffmpeg unreferenced. (Follow-up to #223, which is merged.)
2. **Destroy relies on `SIGTERM` alone.** Escalating to `SIGKILL` after a short grace period is what actually reaps a blocked child. Worth pairing with ending ffmpeg's stdin first, so it can flush and exit on its own rather than being killed — destroying a piped `Readable` does *not* end the destination `Writable`, which is a trap I hit.

If someone here can still reproduce, the useful diagnostic is `ps -o pid,etime,args -C ffmpeg` — the `etime` column immediately separates a genuinely stuck orphan from a recording that is merely in progress.
